src/libAtomVM/ets.c

```diff
@@ -240,71 +243,105 @@ static bool true_pred(struct EtsTable *table, void *data)
     return true;
 }
 
-static void ets_delete_all_tables(struct Ets *ets, GlobalContext *global)
+static void delete_all_tables(struct Ets *ets, GlobalContext *global)
 {
-    ets_delete_tables_internal(ets, true_pred, NULL, global);
+    delete_tables_pred(ets, true_pred, NULL, global);
 }
 
-static EtsErrorCode ets_table_insert(struct EtsTable *ets_table, term entry, Context *ctx)
+static inline bool has_key_at(term tuple, size_t key_index)
 {
-    size_t keypos = ets_table->keypos;
+    return key_index < (size_t) term_get_tuple_arity(tuple);
+}
+
+static struct HNode *tuple_to_insertion_node(struct EtsTable *ets_table, term tuple, GlobalContext *global)
+{
+    bool is_duplicate_bag = ets_table->table_type == EtsTableDuplicateBag;
+    size_t key_index = ets_table->key_index;
+
+    if (is_duplicate_bag) {
+        term key = term_get_tuple_element(tuple, key_index);
+        term old_tuples = ets_hashtable_lookup(ets_table->hashtable, key, global);
+        assert(term_is_list(old_tuples));
+        bool is_new = term_is_nil(old_tuples);
+        return ets_hashtable_new_node_from_list(is_new ? term_nil() : old_tuples, tuple, key_index);
+    }
+    return ets_hashtable_new_node(tuple, key_index);
+}
+
+static struct HNode **list_to_insertion_nodes(struct EtsTable *ets_table, term list, size_t size, GlobalContext *global)
+{
+    size_t last_i = 0;
+    struct HNode **nodes = malloc(size * sizeof(struct HNode *));
+    if (IS_NULL_PTR(nodes)) {
+        goto oom;
+    }
 
-    if ((size_t) term_get_tuple_arity(entry) < keypos + 1) {
+    while (term_is_nonempty_list(list)) {
+        term tuple = term_get_list_head(list);
+        nodes[last_i] = tuple_to_insertion_node(ets_table, tuple, global);
```

Comment:
I did a check here, but I'd love a confirm from your side: in case of OOM, do we remain in a consistent state? I want to be sure we don't leave half inserted stuff or any other incosistency.